src/MatchingResult.php

```diff
@@ -97,18 +97,29 @@ public function route(): Route
 
     public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
     {
-        if (!$this->isSuccess()) {
-            return $handler->handle($request);
+        if ($this->isSuccess() && $this->dispatcher !== null) {
+            return $this
+                ->buildDispatcher($this->dispatcher, $this->route)
+                ->dispatch($request, $handler);
         }
 
-        // Inject dispatcher only if we have not previously injected.
+        return $handler->handle($request);
+    }
+
+    private function buildDispatcher(
+        ?MiddlewareDispatcher $dispatcher,
+        Route $route,
+    ): MiddlewareDispatcher {
+        if ($dispatcher === null) {
+            throw new RuntimeException(sprintf('There is no dispatcher in the route %s.', $route->getData('name')));
+        }
+
+        // Don't add middlewares to dispatcher if we did it earlier.
         // This improves performance in event-loop applications.
-        if ($this->dispatcher !== null && !$this->route->getData('hasDispatcher')) {
-            $this->route->injectDispatcher($this->dispatcher);
+        if ($dispatcher->hasMiddlewares()) {
+            return $dispatcher;
```

Comment:
`Router` pass middlware dispatcher to `MatchingResult` and this dispatcher almost always contain middlewares. In this case route middlewares will not work.

Reply:
I don't understand what you are saying